File: LeetCode/中等/2008. 出租车的最大盈利/solution2008.cpp
/*
url: https://leetcode.cn/problems/maximum-earnings-from-taxi/description/
标签: 动态规划
*/

#include <bits/stdc++.h>

using namespace std;

class Solution2008 {
public:
    // LeetCode上执行用时388ms
    long long maxTaxiEarnings(int n, vector<vector<int>>& rides) {
        vector<long long> dp(n + 1, 0); // dp[i]表示在i位置的最大收益
        vector<vector<vector<int>>> map(n + 1); // map[i]记录终点为i的ride
        for (auto& ride : rides) {
            /* 这里ride是vector<int>&，放进去会发生复制，所以maxTaxiEarnings_implementation2()改了下，
            里面装vector<int>*，效果类似Java
            */
            map[ride[1]].push_back(ride);
        }
        long long ans = 0;
        for (int i = 1; i <= n; i++) {
            // 在i位置有两种情况，刚放下一个人，或者没有放下人，也就是典型的选与不选的分类情况
            dp[i] = dp[i - 1]; // 没放下人
            // 在i位置放下一个人的情况
            for (auto& ride : map[i]) {
                dp[i] = max(dp[i], dp[ride[0]] + ride[1] - ride[0] + ride[2]);
            }
            ans = max(ans, dp[i]);
        }
        return ans;
    }

    // 改成装vector<int>*，效果类似Java，LeetCode上执行用时366ms
    long long maxTaxiEarnings_implementation2(int n, vector<vector<int>>& rides) {
        vector<long long> dp(n + 1, 0); // dp[i]表示在i位置的最大收益
        vector<vector<vector<int>*>> map(n + 1); // map[i]记录终点为i的ride
        for (auto& ride : rides) {
            map[ride[1]].push_back(&ride);
        }
        long long ans = 0;
        for (int i = 1; i <= n; i++) {
            // 在i位置有两种情况，刚放下一个人，或者没有放下人，也就是典型的选与不选的分类情况
            dp[i] = dp[i - 1]; // 没放下人
            // 在i位置放下一个人的情况
            for (auto& ride : map[i]) {
                dp[i] = max(dp[i], dp[(*ride)[0]] + (*ride)[1] - (*ride)[0] + (*ride)[2]);
            }
            ans = max(ans, dp[i]);
        }
        return ans;
    }
};

int main(int argc, char const *argv[]) {
    Solution2008 solu;
    int n = 5;
    vector<vector<int>> rides = {{2,5,4},{1,5,1}};
    cout << solu.maxTaxiEarnings(n, rides) << endl;
    cout << solu.maxTaxiEarnings_implementation2(n, rides) << endl;
    return 0;
}

File: LeetCode/中等/2008. 出租车的最大盈利/Solution2008.java
/*
url: https://leetcode.cn/problems/maximum-earnings-from-taxi/description/
标签: 动态规划
*/

import java.util.*;

public class Solution2008 {
    public long maxTaxiEarnings(int n, int[][] rides) {
        List<List<int[]>> map = new ArrayList<>();
        for (int i = 0; i < n + 1; i++) {
            map.add(new ArrayList<>());
        }
        for (int[] ride : rides) {
            map.get(ride[1]).add(ride);
        }
        long[] dp = new long[n + 1];
        long ans = 0;
        for (int i = 1; i <= n; i++) {
            dp[i] = dp[i - 1]; // 没放下人
            // 刚放下一个人
            for (int[] ride : map.get(i)) {
                dp[i] = Math.max(dp[i], dp[ride[0]] + ride[1] - ride[0] + ride[2]);
            }
            ans = Math.max(ans, dp[i]);
        }
        return ans;
    }

    /* 这个Java代码在LeetCode上运行用时38ms，而C++我的两个实现用时是388ms和366ms，消耗内存也是Java更少，
    LeetCode1466. 重新规划路线官方题解的Java和C++实现对比也是，Java用时似乎有些过于短了，不过看到知乎上的回答:
    <https://www.zhihu.com/question/461560216/answer/2895703314>
    <https://www.zhihu.com/question/50137261/answer/119780292>
    <https://www.zhihu.com/question/528658985/answer/2661467570>
    <https://www.zhihu.com/question/456618030/answer/1856628789>
    大概是LeetCode后台Java的运行用时计算规则相比其它语言有优势，其它语言可能把外围代码的时间算进去了，Java没有。

    此外根据LeetCode上的C++报错信息，LeetCode后台加上了sanitizer检测地址错误等，所以运行也会变慢。
    */

    public static void main(String[] args) {
        Solution2008 solu = new Solution2008();
        System.out.println(solu.maxTaxiEarnings(5, new int[][]{{2,5,4}, {1,5,1}}));
    }
}
